geode-apis-compatible-with-redis/src/integrationTest/java/org/apache/geode/redis/internal/executor/key/AbstractDumpRestoreIntegrationTest.java

```diff
@@ -0,0 +1,264 @@
+
+import org.apache.geode.internal.InternalDataSerializer;
+import org.apache.geode.internal.serialization.DataSerializableFixedID;
+import org.apache.geode.redis.RedisIntegrationTest;
+import org.apache.geode.redis.internal.data.RedisHash;
+import org.apache.geode.redis.internal.data.RedisSet;
+import org.apache.geode.redis.internal.data.RedisSortedSet;
+import org.apache.geode.redis.internal.data.RedisString;
+
+public abstract class AbstractDumpRestoreIntegrationTest implements RedisIntegrationTest {
+
+  private RedisAdvancedClusterCommands<String, String> lettuce;
+  protected JedisCluster jedis;
+  private static String STRING_VALUE;
+  private static byte[] RESTORE_BYTES;
+
+  @BeforeClass
+  public static void setupClass() {
+    InternalDataSerializer.getDSFIDSerializer().registerDSFID(
+        DataSerializableFixedID.REDIS_SET_ID,
+        RedisSet.class);
+    InternalDataSerializer.getDSFIDSerializer().registerDSFID(
+        DataSerializableFixedID.REDIS_STRING_ID,
+        RedisString.class);
+    InternalDataSerializer.getDSFIDSerializer().registerDSFID(
+        DataSerializableFixedID.REDIS_HASH_ID,
+        RedisHash.class);
+    InternalDataSerializer.getDSFIDSerializer().registerDSFID(
+        DataSerializableFixedID.REDIS_SORTED_SET_ID,
+        RedisSortedSet.class);
+  }
+
+  @Before
+  public void setup() {
+    jedis = new JedisCluster(new HostAndPort(BIND_ADDRESS, getPort()), REDIS_CLIENT_TIMEOUT);
+
+    RedisClusterClient client =
+        RedisClusterClient.create(String.format("redis://%s:%d", BIND_ADDRESS, getPort()));
+    lettuce = client.connect().sync();
+
+    STRING_VALUE = "It's a mad, mad, mad, mad, mad world";
+    lettuce.set("set-dump-value", STRING_VALUE);
+    RESTORE_BYTES = lettuce.dump("set-dump-value");
+  }
+
+  @After
+  public void tearDown() {
+    flushAll();
+  }
+
+  @Test
+  public void dumpTakesExactlyOneArgument() {
+    assertExactNumberOfArgs(jedis, Protocol.Command.DUMP, 1);
+  }
+
+  @Test
+  public void restoreErrorsWithUnknownOption() {
+    assertThatThrownBy(
+        () -> jedis.sendCommand("key", Protocol.Command.RESTORE, "key", "0", "", "FOO"))
+            .hasMessageContaining("ERR syntax error");
+  }
+
+  @Test
+  public void restoreFails_whenKeyAlreadyExists() {
+    lettuce.set("restored", "already exists");
+
+    assertThatThrownBy(() -> lettuce.restore("restored", 0, RESTORE_BYTES))
+        .hasMessage("BUSYKEY Target key name already exists.");
+  }
+
+  @Test
+  public void restoreFails_whenTTLisNegative() {
+    assertThatThrownBy(() -> lettuce.restore("restored", -1, RESTORE_BYTES))
+        .hasMessage("ERR Invalid TTL value, must be >= 0");
+  }
+
+  @Test
+  public void restoreFails_whenTTLisNotANumber() {
+    assertThatThrownBy(() -> jedis.sendCommand("restored".getBytes(), Protocol.Command.RESTORE,
+        "restored".getBytes(), "not-an-integer".getBytes(), RESTORE_BYTES))
+            .hasMessage("ERR value is not an integer or out of range");
```

Comment:
Could we use the `stringToBytes` method from `Coder` for these?

Reply:
Since this is all client-side (i.e. how a client might call our API) I'd prefer not to imply that we need to use the `Coder` methods here.

Reply:
I definitely see your point, but I think we decided as a team to switch everything over to use the `stringToBytes` method from Coder.  The only places we don't use that method (other than these dump/restore tests) are when we use a CHARSET other than UTF-8 or a random spot in `RedisString` that must've been overlooked.

Reply:
Marking as resolved per our team conversation.  Thanks for bringing it up!